source/server/async_processing_loop.h

```diff
@@ -0,0 +1,28 @@
+#ifndef NIDEVICE_GRPC_DEVICE_RUN_ASYNC_LOOP_H
+#define NIDEVICE_GRPC_DEVICE_RUN_ASYNC_LOOP_H
+
+#include <grpcpp/grpcpp.h>
+
+namespace nidevice_grpc {
+static void* NO_TAG = nullptr;
+
+class CompletionQueueElement {
+ public:
+  virtual ~CompletionQueueElement() {}
+  virtual void process() = 0;
+};
+
+template <typename TService>
+void run_async_processing_loop(::grpc::ServerCompletionQueue* completion_queue, TService& service)
+{
+  void* tag;
+  bool ok;
+  service.register_async_functions(completion_queue);
```

Comment:
This register call should go in the `DAQService` constructor. Then this wouldn't need to be a template. I plan to do this for this change , just putting off that kind of housekeeping until the direction is set.

Reply:
Actually, that doesn't work. You can't do that before the service is registered/started.

Leaving as is.